opengovsg/plumber
# PLU-125: [EXCEL-6] Dynamic data for excel

## Problem
We need to provide a nice UX for users to select appropriate files, worksheets, etc.

## Solution
We will use dynamic data dropdowns to provide a selection to the user.

The tricky bit is that the dynamic data has to be dependent on other dynamic data fields (_e.g. only query worksheets after selecting the excel file_). It turns out automatisch already had this, just implemented in a non-conventional manner (it added non-directly-used variables to a `useEffect` dependency list). We broke it when we enabled react linting.

This PR fixes this dependent dynamic data regression, and adds dynamic data for excel files, worksheets and tables in the excel file.

**Files dynamic data**
<img width="500" src="https://github.com/opengovsg/plumber/assets/135598754/8e62993d-bc9f-4891-8f82-621385ffc905" />

**Worksheets dynamic data**
<img width="500" src="https://github.com/opengovsg/plumber/assets/135598754/c38f203d-cb09-487c-8539-2c32ef2870f5" />

**Tables dynamic data**
<img width="500" src="https://github.com/opengovsg/plumber/assets/135598754/4c3d685c-08c6-4645-9f0d-a0c7fdfd0f4d" />

**Worksheets in different file**
<img width="500" src="https://github.com/opengovsg/plumber/assets/135598754/5b897fae-8f0f-4eca-851e-213075f252dc" />

**Empty results in dependent fields if no file selected**
<img width="500" src="https://github.com/opengovsg/plumber/assets/135598754/d895fa75-77b1-46da-b06d-7516745b43a1" />

## Tests
- Check that only excel files show up in the file drop down.
- Check that only sheets from the currently selected excel file show up in the sheets drop down. If no file is selected, no results are returned.
- Check that only tables from the currently selected excel file show up in the tables drop down. If no file is selected, no results are returned.
- Regression test: check that dynamic data is not broken for other apps (e.g. telegram, vault workspace)
